text/0013-prepare-blockbuilder-and-core-runtime-apis-for-mbms.md

```diff
@@ -0,0 +1,113 @@
+# RFC-0013: Prepare `Core` runtime API for MBMs
+
+|                 |                                                                             |
+| --------------- | --------------------------------------------------------------------------- |
+| **Start Date**  | July 24, 2023 |
+| **Description** | Prepare the `Core` Runtime API for Multi-Block-Migrations |
+| **Authors**     | Oliver Tale-Yazdi |
+
+## Summary
+
+Introduces breaking changes to the `Core` runtime API by letting `Core::initialize_block` return an enum. The versions of `Core` is bumped from 4 to 5.
+
+## Motivation
+
+The main feature that motivates this RFC are Multi-Block-Migrations (MBM); these make it possible to split a migration over multiple blocks.  
+Further it would be nice to not hinder the possibility of implementing a new hook `poll`, that runs at the beginning of the block when there are no MBMs and has access to `AllPalletsWithSystem`. This hook can then be used to replace the use of `on_initialize` and `on_finalize` for non-deadline critical logic.  
+In a similar fashion, it should not hinder the future addition of a `System::PostInherents` callback that always runs after all inherents were applied.
+
+## Stakeholders
+
+- Substrate Maintainers: They have to implement this, including tests, audit and
+  maintenance burden.
+- Polkadot Runtime developers: They will have to adapt the runtime files to this breaking change.
+- Polkadot Parachain Teams: They have to adapt to the breaking changes but then eventually have
+  multi-block migrations available.
+
+## Explanation
+
+
+### `Core::initialize_block`
+
+This runtime API function is changed from returning `()` to `ExtrinsicInclusionMode`:
+
+```patch
+fn initialize_block(header: &<Block as BlockT>::Header)
++  -> ExtrinsicInclusionMode;
+```
+
+With `ExtrinsicInclusionMode` is defined as:
+
+```rust
+enum ExtrinsicInclusionMode {
+  /// All extrinsics are allowed in this block.
+  AllExtrinsics,
+  /// Only inherents are allowed in this block.
+  OnlyInherents,
+}
+```
+
+A block author MUST respect the `ExtrinsicInclusionMode` that is returned by `initialize_block`. The runtime MUST reject blocks that have extrinsics in them while `OnlyInherents` was returned.
```

Comment:
```suggestion
A block author MUST respect the `ExtrinsicInclusionMode` that is returned by `initialize_block`. The runtime MUST reject blocks that have non-inherent extrinsics in them while `OnlyInherents` was returned.
```

nit, or else we have got the definition wrong here: https://paritytech.github.io/polkadot-sdk/master/sp_runtime/generic/struct.UncheckedExtrinsic.html

Although these namings probably need revision based on https://github.com/paritytech/polkadot-sdk/pull/2280 as well

Reply:
Oh thanks, yea this was def wrong before.